### PR TITLE
Feature ETP-2197: Improve FK handling, backup management and logging in partition operations

### PR DESCRIPTION
**Pull request by RomanMagnoli**

### CODE REVIEW

File: src-util/modulescript/src/com/etendoerp/db/extended/utils/TableDefinitionComparator.java [hunk 1]
```diff
@@ -62,6 +62,13 @@
    * @throws Exception if any error occurs while parsing XML or querying the database
    */
   public boolean isTableDefinitionChanged(String tableName, ConnectionProvider cp, List<File> xmlFiles) throws Exception {
+    // If no XML files are provided, consider the definition changed so that
+    // installations that introduce new XML files will be processed by the ModuleScript.
+    if (xmlFiles == null || xmlFiles.isEmpty()) {
+      logger.debug("No XML files for table {}: treating as changed to allow creation/apply during install.", tableName);
```
Comment (etendobot) [inline src-util/modulescript/src/com/etendoerp/db/extended/utils/TableDefinitionComparator.java:68]: **Suggestion**

Consider ensuring the debug log message provides enough context for troubleshooting, especially in multi-table operations (e.g., include module or file path if relevant).

```java
if (xmlFiles == null || xmlFiles.isEmpty()) {
      logger.debug("No XML files for table {}: treating as changed to allow creation/apply during install.", tableName);
      return true;
    }
```

**Rationale**: While this is not required, improving log context can help diagnostics in complex Etendo module installations. Current line is fine, but more verbose logs (table+file/module) are often helpful per Etendo installation troubleshooting docs.

File: src-util/modulescript/src/com/etendoerp/db/extended/utils/TableDefinitionComparator.java [hunk 2]
```diff
@@ -75,6 +82,41 @@
     return !xmlColumns.equals(dbColumns);
   }
 
+  /**
+   * Returns columns to add and to remove to make DB match XML.
+   * added -> columns present in XML but missing in DB
+   * removed -> columns present in DB but missing in XML
+   */
+  public ColumnDiff diffTableDefinition(String tableName, ConnectionProvider cp, List<File> xmlFiles) throws Exception {
+    Map<String, ColumnDefinition> xmlColumns = new LinkedHashMap<>();
```
Comment (etendobot) [inline src-util/modulescript/src/com/etendoerp/db/extended/utils/TableDefinitionComparator.java:91]: **Suggestion**

Consider adding a short Javadoc description for the public diffTableDefinition method, similar to other public APIs in Etendo utility classes.

```java
public ColumnDiff diffTableDefinition(String tableName, ConnectionProvider cp, List<File> xmlFiles) throws Exception {
```

**Rationale**: Consistent method documentation helps maintainability and user understanding, as seen in comparable Etendo source.

File: src-util/modulescript/src/com/etendoerp/db/extended/utils/TableDefinitionComparator.java [hunk 3]
```diff
@@ -221,6 +263,23 @@
       return name;
     }
 
+    // New getters for other properties to allow external callers to build SQL
+    public String getDataType() {
+      return dataType;
```
Comment (etendobot) [inline src-util/modulescript/src/com/etendoerp/db/extended/utils/TableDefinitionComparator.java:268]: **Suggestion**

For public getter methods, ensure JavaDocs or comments indicate the meaning and possible nullability/defaults, to align with Etendo's standard on property accessors.

```java
public String getDataType() {
      return dataType;
    }

    public Integer getLength() {
      return length;
    }

    public Boolean isNullable() {
      return isNullable;
    }

    public Boolean isPrimaryKey() {
      return isPrimaryKey;
    }
```

**Rationale**: Clear documentation of model properties reduces misuse by API users and is standard in Etendo entity and utility classes.

File: src-util/buildvalidation/src/com/etendoerp/db/extended/buildvalidation/CreateExcludeFilter.java [hunk 4]
```diff
@@ -267,12 +274,16 @@ public Set<String> findAllForeignKeysReferencing(String targetTable) {
...
-   * @param xmlFiles the list of XML files to inspect
+   * @param xmlFiles
+   *     the list of XML files to inspect
    * @return the name of the primary key if found, or null if missing or on error
    */
   public static String findPrimaryKey(List<File> xmlFiles) {
     try {
       for (File xml : xmlFiles) {
+        if (StringUtils.contains(xml.getAbsolutePath(), "modifiedTables")) {
+          continue;
```
Comment (etendobot) [inline src-util/buildvalidation/src/com/etendoerp/db/extended/buildvalidation/CreateExcludeFilter.java:285]: **Suggestion**

Consider documenting within the code why modifiedTables are skipped when searching for primary keys, referencing Etendo's practice for model XML traversal.

```java
if (StringUtils.contains(xml.getAbsolutePath(), "modifiedTables")) {
```

**Rationale**: Etendo's customizations to XML model file handling sometimes require excluding certain directories or files. Explicit commentary can ease maintenance and clarify the rationale for team members unfamiliar with this particular skip condition.

File: tool/unpartition.py [hunk 5]
```diff
@@ -271,38 +351,234 @@
...
+
+        # Determine columns to copy (exclude any etarc_ columns)
+        cur.execute(r"""
+            SELECT column_name FROM information_schema.columns
+            WHERE table_schema = %s AND table_name = %s AND (column_name NOT LIKE %s AND column_name NOT LIKE %s)
+            ORDER BY ordinal_position
+        """, (schema, table_name, ETARC_PATTERN, ETRAC_PATTERN))
+        cols_to_copy = [r[0] for r in cur.fetchall()]
+        if not cols_to_copy:
+            raise Exception(f"No columns found to copy for table {schema}.{table_name} after removing etarc_ columns")
```
Comment (etendobot) [inline tool/unpartition.py:511]: **Suggestion**

Consider an early return with an explanatory message instead of raising a generic Exception, which may help keep the workflow intact and allow other tables in batch mode to proceed.

```python
if not cols_to_copy:
    print_message(f"No columns found to copy for table {schema}.{table_name} after removing etarc_ columns", "ERROR")
    return
```

**Rationale**: Graceful failure in scripts is typically recommended to avoid stopping batch operations entirely due to issues in one table.

File: src-util/modulescript/src/com/etendoerp/db/extended/utils/TableDefinitionComparator.java [hunk 6]
```diff
@@ -221,6 +263,23 @@
...
+    }
+
+    public Integer getLength() {
+      return length;
+    }
+
+    public Boolean isNullable() {
+      return isNullable;
```
Comment (etendobot) [inline src-util/modulescript/src/com/etendoerp/db/extended/utils/TableDefinitionComparator.java:276]: **Suggestion**

Consider also providing 'getIsNullable()' and 'getIsPrimaryKey()' in addition to 'isNullable()' and 'isPrimaryKey()' for consistency with standard JavaBean patterns, ensuring interoperability with frameworks that expect getter naming conventions.

```java
public Boolean isNullable() {
      return isNullable;
    }

    public Boolean getIsNullable() {
      return isNullable;
    }

    public Boolean isPrimaryKey() {
      return isPrimaryKey;
    }

    public Boolean getIsPrimaryKey() {
      return isPrimaryKey;
    }
```

**Rationale**: Etendo and JavaBean conventions often require both 'isX' and 'getX' variants for boolean properties to maximize compatibility with tools and frameworks that rely on reflection.

**Review by etendobot — CHANGES_REQUESTED** [review]
Changes requested by agent. Please resolve blocking issues.

File: tool/unpartition.py [hunk 7]
```diff
@@ -67,6 +72,7 @@
             color = Style.RED
         print(f"{color}{emoji} {table:<25} → {status:<8} | {message}{Style.RESET}")
 
+    print_message("Execute './gradlew update.database -Dforce=yes' to complete the operation", "HEADER")
```
Comment (etendobot) [inline tool/unpartition.py:76]: **Suggestion**

Clarify that './gradlew update.database -Dforce=yes' is required to re-apply constraints after unpartitioning.

```python
+    print_message("Once the operation is completed, execute './gradlew update.database -Dforce=yes' to re-apply constraints.", "HEADER")
```

**Rationale**: While users familiar with Etendo will know this, being explicit improves usability for less-experienced users. Not critical, but would make the summary step clearer.

File: src-util/modulescript/src/com/etendoerp/db/extended/modulescript/PartitionedConstraintsHandling.java [hunk 8]
```diff
@@ -522,13 +1038,45 @@
...
-                sql.append(String.format(updateColumnSQL, relatedTableName, partitionColumn,
-                    partitionField, tableName, pkField, relatedTableName, relationColumn,
-                    relatedTableName, partitionColumn));
-                sql.append(
-                    String.format(addPartitionedForeignKeySQL, relatedTableName, foreignKey,
-                        relationColumn, partitionColumn, tableName, pkField, partitionField));
+                // If both the helper column and the FK constraint already exist on the child table,
+                // skip creating them — they are assumed correct.
+                boolean colExists = false;
+                boolean fkExists = false;
```
Comment (etendobot) [inline src-util/modulescript/src/com/etendoerp/db/extended/modulescript/PartitionedConstraintsHandling.java:1044]: **Suggestion**

Optional: Consider adding unit/integration tests for the new existence checks for partition columns and FK constraints to prevent regressions as logic becomes more complex.

```java
    // If both the helper column and the FK constraint already exist on the child table,
    // skip creating them — they are assumed correct.
    boolean colExists = false;
    boolean fkExists = false;
    try {
      colExists = columnExists(cp, relatedTableName, partitionColumn);
    } catch (Exception e) {
      log4j.warn("Could not check existence of column {} on {}: {}", partitionColumn, relatedTableName,
          e.getMessage());
    }
    try {
      fkExists = constraintExists(cp, relatedTableName, foreignKey);
    } catch (Exception e) {
      log4j.warn("Could not check existence of constraint {} on {}: {}", foreignKey, relatedTableName,
          e.getMessage());
    }
```

**Rationale**: While the new checks for helper column and FK existence are robust, Etendo best practices encourage regression tests for data definition operations where possible.

File: src-util/modulescript/src/com/etendoerp/db/extended/modulescript/PartitionedConstraintsHandling.java [hunk 9]
```diff
@@ -164,20 +326,395 @@
...
+    }
+  }
+
+  /**
+   * Returns the last processed timestamp for the given table, or null if none.
+   */
+  private java.sql.Timestamp getLastProcessed(ConnectionProvider cp, String tableName) {
+    String sql = "SELECT last_processed FROM " + BACKUP_SCHEMA + ".processing_metadata WHERE lower(table_name) = lower(?)";
+    try (PreparedStatement ps = cp.getPreparedStatement(sql)) {
+      ps.setString(1, tableName);
```
Comment (etendobot) [inline src-util/modulescript/src/com/etendoerp/db/extended/modulescript/PartitionedConstraintsHandling.java:466]: **Suggestion**

Consider explicitly logging the SQL statement and exception on failure inside executeConstraintSqlIfNeeded, not just in the calling method.

```java
    try (PreparedStatement ps = cp.getPreparedStatement(sql)) {
      ps.executeUpdate();
    } catch (Exception e) {
      log4j.error("Error executing SQL: {}", sql, e);
      throw e;
    }
```

**Rationale**: Explicit error logging including the SQL statement helps with traceability and debugging during complex schema migrations, as recommended in Etendo migration troubleshooting patterns.

File: tool/migrate.py [hunk 10]
```diff
@@ -1341,7 +1880,7 @@
     return psycopg2.connect(**params)
 
 def apply_session_config(conn, properties, dry_run):
-    session_config = properties.get('bbdd.sessionConfig')
+    session_config = properties.get(BBDD_SESSIONCONFIG)
```
Comment (etendobot) [inline tool/migrate.py:1883]: **Suggestion**

Consider maintaining consistent usage of the extracted constant BBDD_SESSIONCONFIG throughout the codebase to avoid divergence. Also, verify that other places where 'bbdd.sessionConfig' as a string is used are similarly updated.

```python
def apply_session_config(conn, properties, dry_run):
    session_config = properties.get(BBDD_SESSIONCONFIG)
```

**Rationale**: Promotes consistency across the code and makes future refactors less error-prone.

File: src-test/src/com/etendoerp/db/extended/handler/PartitionTableEventHandlerTest.java [hunk 11]
```diff
@@ -91,15 +103,45 @@ void setUp() throws Exception {
...
     when(deleteEvent.getTargetInstance()).thenReturn(tableConfig);
     when(tableConfig.getTable()).thenReturn(table);
     when(table.getDBTableName()).thenReturn("mock_table");
     when(tableConfig.getEntity()).thenReturn(tableConfigEntity);
+    when(tableConfig.getColumn()).thenReturn(column);
+    when(column.getDBColumnName()).thenReturn("updated");
   }
 
+  private void setupMockJdbc(String hasNullsValue, String hasUniqueValue) throws Exception {
+    OBDal obDal = mock(OBDal.class);
```
Comment (etendobot) [inline src-test/src/com/etendoerp/db/extended/handler/PartitionTableEventHandlerTest.java:118]: **Suggestion**

Consider adding Javadoc comments describing both parameters and their expected values for this new overload for future readability.

```java
  private void setupMockJdbc(String hasNullsValue, String hasUniqueValue) throws Exception {
```

**Rationale**: Documenting test utility methods improves maintainability and clarity for other contributors, aligning with Etendo test best practices.

File: src-util/buildvalidation/src/com/etendoerp/db/extended/buildvalidation/CreateExcludeFilter.java (hunk 4 above)
Comment (etendobot) [inline src-util/buildvalidation/src/com/etendoerp/db/extended/buildvalidation/CreateExcludeFilter.java:285]: **Suggestion**

The exclusion of XML files in 'modifiedTables' for primary key search is reasonable if those files are known to not contain base metadata. Ensure this is consistent with Etendo's table schema evolution patterns and that no PKs might be present in 'modifiedTables' unexpectedly.

```java
if (StringUtils.contains(xml.getAbsolutePath(), "modifiedTables")) {
```

**Rationale**: Modified tables often should not override PK definitions; this aligns with modular model update practices found in Etendo code, but developers should be aware of edge-case schema evolutions.

File: src-util/modulescript/src/com/etendoerp/db/extended/utils/TableDefinitionComparator.java (hunk 6 above)
Comment (etendobot) [inline src-util/modulescript/src/com/etendoerp/db/extended/utils/TableDefinitionComparator.java:276]: **Suggestion**

Consider using primitive 'boolean' instead of 'Boolean' for getter return types unless null is a valid state for these properties.

```java
public Boolean isNullable() {
      return isNullable;
    } // (no change required)
```

**Rationale**: Using Boolean may introduce the risk of NPE in client code if internal properties aren't explicitly set; primitive types can avoid this.

File: src/com/etendoerp/db/extended/handler/PartitionTableEventHandler.java [hunk 12]
```diff
@@ -30,15 +38,44 @@
...
-import java.sql.SQLException;
+import com.etendoerp.db.extended.data.TableConfig;
 
 public class PartitionTableEventHandler extends EntityPersistenceEventObserver {
   private static final Logger logger = LogManager.getLogger();
   private static final Entity[] entities = { ModelProvider.getInstance().getEntity(TableConfig.ENTITY_NAME) };
+  public static final String ETARC_COULD_NOT_RETRIEVE_TABLES = "ETARC_CouldNotRetrieveTables";
+
+  private static final Pattern IDENTIFIER = Pattern.compile("^[A-Za-z_][A-Za-z0-9_]*$");
+  private static final String DEFAULT_SCHEMA = "public";
```
Comment (etendobot) [inline src/com/etendoerp/db/extended/handler/PartitionTableEventHandler.java:49]: **Suggestion**

Consider documenting the regex intent via inline comment or JavaDoc for maintainability.

```java
private static final Pattern IDENTIFIER = Pattern.compile("^[A-Za-z_][A-Za-z0-9_]*$"); // consider JavaDoc for pattern intent
```

**Rationale**: Providing an explicit comment on identifier patterns helps new contributors understand the restrictions (esp. if supporting different DBs in future).

File: tool/migrate.py [hunk 13]
```diff
@@ -224,10 +236,118 @@
...
+    except (psycopg2.Error, psycopg2.DatabaseError) as e:
+        print_message(f"  - Table type: ERROR - {e}", "DEBUG")
+        # If there's an error, rollback to avoid affecting main transaction
+        try:
+            conn.rollback()
+        except psycopg2.Error:
+            pass
+
+def debug_table_status(conn, table_name, schema, step_description):
+    """Debug function to check table existence and properties during migration"""
```
Comment (etendobot) [inline tool/migrate.py:324]: **Suggestion**

Consider using context managers (with conn: ...) in all points where a rollback is made in the exception handler. While it is not strictly required for rollback cases, it can help ensure that the connection's transactional state is safely managed and errors are consistently handled. No change required, this is a style/best-practices suggestion.

```python
def debug_table_status(conn, table_name, schema, step_description):
    """Debug function to check table existence and properties during migration"""
    try:
        print_message(f"DEBUG [{step_description}] Checking table status for {schema}.{table_name}:", "DEBUG")

        # Check existence
        exists = table_exists(conn, table_name, schema)
        print_message(f"  - Table exists: {exists is not None} (in schema: {exists})", "DEBUG")

        if exists:
            # Check if partitioned
            is_part = is_table_partitioned(conn, table_name, schema)
            print_message(f"  - Is partitioned: {is_part}", "DEBUG")

            # Get row count - use a separate connection/cursor to avoid affecting main transaction
            _get_table_row_count(conn, schema, table_name)

            # Check table type from pg_class - simplified to avoid transaction issues
            _get_table_type_info(conn, schema, table_name)

    except (psycopg2.Error, psycopg2.DatabaseError, ValueError) as e:
        print_message(f"DEBUG [{step_description}] Error checking table status: {e}", "ERROR")
        # If there's an error, rollback to avoid affecting main transaction
        try:
            conn.rollback()
        except psycopg2.Error:
            pass
```

**Rationale**: This is for improved readability and maintenance; Etendo standards allow flexible connection management but consistent usage of context-managers helps reinforce intent.

**Review by etendobot — APPROVED** [review]

File: src/com/etendoerp/db/extended/handler/PartitionTableEventHandler.java [hunk 14]
```diff
@@ -47,6 +84,39 @@
...
+          "SELECT CASE WHEN EXISTS (SELECT 1 FROM " + schema + "." + table +
+              " WHERE " + column + " IS NULL) THEN 'Y' ELSE 'N' END AS hasnull";
+
+      String hasNulls = "N";
+      try (Connection connection = OBDal.getInstance().getConnection(false);
+           PreparedStatement ps = connection.prepareStatement(nullCheckSql);
+           ResultSet rs = ps.executeQuery()) {
+        if (rs.next()) {
+          hasNulls = rs.getString("hasnull");
+        }
```
Comment (etendobot) [inline src/com/etendoerp/db/extended/handler/PartitionTableEventHandler.java:107]: **Suggestion**

While current usage appears safe, consider always using Objects.equals or explicit null checks in complex logic for comparison involving possible nulls.

```java
hasNulls = rs.getString("hasnull"); // Defensive: consider Objects.equals for null safety if used in larger context
```

**Rationale**: Consistent defensive checks for null protect against future accidental NPEs as logic evolves.

File: src-util/modulescript/src/com/etendoerp/db/extended/modulescript/PartitionedConstraintsHandling.java [hunk 15]
```diff
@@ -17,47 +17,53 @@
...
-  private static final String[] moduleDirs = new String[] {MODULES_BASE, MODULES_CORE, MODULES_JAR};
   public static final String SEPARATOR = "=======================================================";
+  private static final String SRC_DB_DATABASE_MODEL_TABLES = "src-db/database/model/tables";
+  private static final String SRC_DB_DATABASE_MODEL_MODIFIED_TABLES = "src-db/database/model/modifiedTables";
+  private static final Logger log4j = LogManager.getLogger();
+  private static final String[] moduleDirs = new String[]{ MODULES_BASE, MODULES_CORE, MODULES_JAR };
+  // Backup infrastructure
+  private static final String BACKUP_SCHEMA = "etarc_backups";
+  private static final String BACKUP_METADATA_TABLE = "backup_metadata"; // in BACKUP_SCHEMA
+  private static final int MAX_BACKUPS_PER_TABLE = 5;
```
Comment (etendobot) [inline src-util/modulescript/src/com/etendoerp/db/extended/modulescript/PartitionedConstraintsHandling.java:67]: **Suggestion**

Consider explicitly documenting in code or module README the rationale for limiting backups to 5. While this is a safe default, making the rotation policy (and restore procedures) visible improves maintainability.

```java
private static final int MAX_BACKUPS_PER_TABLE = 5; // Etendo standard: backup rotation limit for partitioned tables
```

**Rationale**: Explicitly documenting critical constants related to data retention helps future maintainers and aligns with Etendo best practices for module extensibility.

File: src-util/modulescript/src/com/etendoerp/db/extended/modulescript/PartitionedConstraintsHandling.java [hunk 16]
```diff
@@ -164,20 +326,395 @@
...
+          }
+        }
+      }
+    } catch (Exception e) {
+      log4j.warn("Failed to cleanup excess backups for table {}: {}", tableName, e.getMessage());
+    }
+  }
+
+  private String createTableBackup(ConnectionProvider cp, String tableName) throws Exception {
+    // Backup strategy: create a new table in BACKUP_SCHEMA with timestamped name and copy data
```
Comment (etendobot) [inline src-util/modulescript/src/com/etendoerp/db/extended/modulescript/PartitionedConstraintsHandling.java:617]: **Suggestion**

Use a formatted timestamp (e.g. yyyyMMddHHmmss) for backup table naming, rather than raw milliseconds, to improve auditability and manual restoration operations.

```java
private String createTableBackup(ConnectionProvider cp, String tableName) throws Exception {
    // Backup strategy: create a new table in BACKUP_SCHEMA with timestamped name and copy data.
    // Consider using a formatted timestamp (yyyyMMddHHmmss or similar) for improved backup tracking.
```

**Rationale**: Using a human-readable timestamp in backup naming aids in manual troubleshooting, especially during DB disaster recovery. This is a non-blocking best-practice refinement.

File: tool/migrate.py [hunk 17]
```diff
@@ -224,10 +236,118 @@
...
+            print_message(f"  - Is partitioned: {is_part}", "DEBUG")
+
+            # Get row count - use a separate connection/cursor to avoid affecting main transaction
+            _get_table_row_count(conn, schema, table_name)
+
+            # Check table type from pg_class - simplified to avoid transaction issues
+            _get_table_type_info(conn, schema, table_name)
+
+    except (psycopg2.Error, psycopg2.DatabaseError, ValueError) as e:
+        print_message(f"DEBUG [{step_description}] Error checking table status: {e}", "ERROR")
```
Comment (etendobot) [inline tool/migrate.py:344]: **Suggestion**

To ensure all exceptions are properly logged and possible rollbacks attempted, consider wrapping all high-level migration logic in a broader Exception to catch unexpected failures. Add a generic except Exception as e: with full logging in addition to the specific exception tuples.

```python
except (psycopg2.Error, psycopg2.DatabaseError, ValueError) as e:
```

**Rationale**: Catching all exceptions at the process entrypoint level with a clear error log and safe rollback is a pattern observed in robust migration tools in the Etendo knowledge base, ensuring unanticipated failures don't leave the database in an inconsistent state.

File: tool/migrate.py [hunk 18]
```diff
@@ -931,250 +1125,594 @@
...
+                debug_table_status(conn, new_partitioned_table_name, actual_schema, "After Data Copy - New Table")
+
+        except (psycopg2.Error, psycopg2.DatabaseError) as e:
+            print_message(f"ERROR: Failed to copy data: {e}", "ERROR")
+            raise
+
+    print_message(f"\n--- Data migration to {actual_schema}.{new_partitioned_table_name} complete. ---", "SUCCESS")
+
+def _print_dropped_constraints_info(dropped_referencing_fks_info, dropped_on_table_fks_info):
+    """Print information about dropped foreign key constraints."""
```
Comment (etendobot) [inline tool/migrate.py:1556]: **Suggestion**

Adding parameter type hints improves function documentation and helps static analysis tools understand usage, especially helpful when manipulating constraint metadata.

```python
def _print_dropped_constraints_info(dropped_referencing_fks_info: list, dropped_on_table_fks_info: list):
```

**Rationale**: Adhering to Python typing best practices in modern Etendo scripts as established in Python code samples in the Wiki and core tools.

**Review by etendobot — APPROVED** [review]

File: src-util/buildvalidation/src/com/etendoerp/db/extended/buildvalidation/CreateExcludeFilter.java (hunk 4 above)
Comment (etendobot) [inline src-util/buildvalidation/src/com/etendoerp/db/extended/buildvalidation/CreateExcludeFilter.java:285]: **Suggestion**

This line helps exclude modifiedTables from PK search, which is an Etendo convention and improves correctness. Consider clarifying the rationale with an inline comment for future maintainers.

```java
if (StringUtils.contains(xml.getAbsolutePath(), "modifiedTables")) {
```

**Rationale**: Making the reasoning explicit improves maintainability for developers unfamiliar with Etendo's directory structure conventions.

File: src/com/etendoerp/db/extended/handler/PartitionTableEventHandler.java [hunk 19]
```diff
@@ -30,15 +38,44 @@
...
+   * }</pre>
+   *
+   * @param id
+   *     the identifier to validate and quote; must not be {@code null}
+   * @return the quoted identifier with internal double quotes escaped
+   * @throws OBException
+   *     if {@code id} is {@code null} or does not match the allowed pattern
+   */
+  private static String quoteIdent(String id) {
+    if (id == null || !IDENTIFIER.matcher(id).matches()) {
```
Comment (etendobot) [inline src/com/etendoerp/db/extended/handler/PartitionTableEventHandler.java:74]: **Suggestion**

Consider adding unit tests for quoteIdent to verify behavior with edge-case inputs, as identifier quoting correctness is crucial.

```java
private static String quoteIdent(String id) { // Consider adding a unit test to ensure expected escaping behavior
```

**Rationale**: Testing this utility helps prevent SQL injection and naming errors as the method is fundamental for safe dynamic SQL and follows Etendo PostgreSQL compatibility patterns.

File: src/com/etendoerp/db/extended/handler/PartitionTableEventHandler.java [hunk 20]
```diff
@@ -121,10 +189,11 @@
 
   private static void logSQLError(SQLException e) {
     logger.error("Error executing SQL query. Message: {}, SQLState: {}, ErrorCode: {}",
-            e.getMessage(), e.getSQLState(), e.getErrorCode());
+        e.getMessage(), e.getSQLState(), e.getErrorCode());
```
Comment (etendobot) [inline src/com/etendoerp/db/extended/handler/PartitionTableEventHandler.java:192]: **Suggestion**

Log the entire SQLException (including stack trace) by passing the exception object as the last parameter to logger.error.

```java
logger.error("Error executing SQL query. Message: {}, SQLState: {}, ErrorCode: {}",
        e.getMessage(), e.getSQLState(), e.getErrorCode(), e);
```

**Rationale**: Including the full exception provides stack trace details, improving debugging following Etendo's error logging best practices.

**Review by Matias-Bernal — APPROVED** [review]

**Review by etendobot — APPROVED** [review]

File: src-util/modulescript/src/com/etendoerp/db/extended/utils/TableDefinitionComparator.java [hunk 21]
```diff
@@ -221,11 +329,27 @@ public String getName() {
...
+      return isPrimaryKey;
+    }
+
     @Override
     public boolean equals(Object o) {
       if (this == o) return true;
-      if (!(o instanceof ColumnDefinition)) return false;
-      ColumnDefinition that = (ColumnDefinition) o;
+      if (!(o instanceof ColumnDefinition that)) return false;
       return Objects.equals(name, that.name);
```
Comment (etendobot) [inline src-util/modulescript/src/com/etendoerp/db/extended/utils/TableDefinitionComparator.java:353]: **Suggestion**

The use of pattern matching for 'instanceof' is a Java language improvement (introduced in Java 16+) and is preferred where supported. Confirm project-wide compatibility with Java 16+ for usage.

```java
if (!(o instanceof ColumnDefinition that)) return false;
```

**Rationale**: Using pattern matching for instanceof increases code conciseness and clarity, aligning with modern Java best practices as seen in other Etendo code base modules. If the project is not yet requiring Java 16 or higher, this change could cause build issues on older environments.
